File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/ColumnToFoundationMove.java
package org.silnith.game.solitaire.move;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import org.silnith.deck.Card;
import org.silnith.deck.Suit;
import org.silnith.game.solitaire.Board;
import org.silnith.game.solitaire.Column;
import org.silnith.game.util.Pair;


/**
 * A move that takes a face-up card from a column and puts it into the
 * Foundation.
 */
public class ColumnToFoundationMove implements SolitaireMove {
	
	/**
	 * Finds all column to foundation moves for a given board.
	 * 
	 * <p>This could potentially find a move for each suit.</p>
	 * 
	 * @param board the board to examine
	 * @return a collection of moves
	 */
	public static Collection<ColumnToFoundationMove> findMoves(final Board board) {
		final Collection<ColumnToFoundationMove> moves = new ArrayList<ColumnToFoundationMove>(Suit.values().length);
		final List<Column> columns = board.getColumns();
		for (int i = 0; i < columns.size(); i++) {
			final Column column = columns.get(i);
			if (column.hasFaceUpCards()) {
				final Card card = column.getTopCard();
				if (board.canAddToFoundation(card)) {
					moves.add(new ColumnToFoundationMove(i, card));
				}
			}
		}
		return moves;
	}
    
    /**
     * The index in the board of the column from which the card is taken.
     */
    private final int sourceColumnIndex;
    
    /**
     * The card being moved.
     */
    private final Card card;
    
    /**
     * Creates a new move that takes a face-up card from a column and puts it
     * into the Foundation.
     * 
     * @param sourceColumnIndex the index into the board of the column from which the
     *         card is taken
     * @param card the card being moved
     */
    public ColumnToFoundationMove(final int sourceColumnIndex, final Card card) {
        super();
        this.sourceColumnIndex = sourceColumnIndex;
        this.card = card;
    }
    
    /**
     * Creates a new move that takes a face-up card from a column and puts it
     * into the foundation.
     * 
     * @param sourceColumnIndex the index into the board of the column from which the
     *         card is taken
     * @param board the board containing the card to move
     * @throws IndexOutOfBoundsException if the source column is out of range
     * @throws IllegalArgumentException if the source column has no card to move
     */
    public ColumnToFoundationMove(final int sourceColumnIndex, final Board board) {
    	this(sourceColumnIndex, board.getColumn(sourceColumnIndex).getTopCard());
    }
    
    /**
     * Returns the card being moved.
     * 
     * @return the moved card
     * @see #getCards()
     */
    public Card getCard() {
        return card;
    }
    
    @Override
    public boolean hasCards() {
        return true;
    }
    
    @Override
    public List<Card> getCards() {
        return Collections.singletonList(card);
    }

    @Override
    public boolean isStockPileModification() {
        return false;
    }

    @Override
    public boolean isStockPileAdvance() {
        return false;
    }

    @Override
    public boolean isStockPileRecycle() {
        return false;
    }

    @Override
    public boolean isFromStockPile() {
        return false;
    }

    @Override
    public boolean isFromFoundation() {
        return false;
    }

    @Override
    public boolean isFromColumn() {
        return true;
    }

    @Override
    public boolean isFromColumn(int columnIndex) {
        return columnIndex == sourceColumnIndex;
    }

    @Override
    public int getSourceColumnIndex() {
        return sourceColumnIndex;
    }

    @Override
    public boolean isToFoundation() {
        return true;
    }

    @Override
    public boolean isToColumn() {
        return false;
    }

    @Override
    public boolean isToColumn(int columnIndex) {
        return false;
    }

	@Override
    public int getDestinationColumnIndex() {
        throw new IllegalStateException("Not a move to a column.");
    }

    @Override
    public Board apply(final Board board) {
        final List<Column> columns = board.getColumns();
		final Column column = columns.get(sourceColumnIndex);
		final Pair<Card, Column> pair = column.extractCard();
		final Card card = pair.getFirst();
		final Column newColumn = pair.getSecond();
		
		final List<Column> newColumns = new ArrayList<>(columns);
		newColumns.set(sourceColumnIndex, newColumn);
		
		final List<Card> stockPile = board.getStockPile();
		final int stockPileIndex = board.getStockPileIndex();
		final Map<Suit, List<Card>> newFoundation = board.getFoundationPlusCard(card);
		
		return new Board(newColumns, stockPile, stockPileIndex, newFoundation);
    }
    
    @Override
    public int hashCode() {
        return Integer.rotateLeft(sourceColumnIndex, 16) ^ card.hashCode();
    }
    
    @Override
    public boolean equals(final Object obj) {
        if (obj instanceof ColumnToFoundationMove) {
            final ColumnToFoundationMove move = (ColumnToFoundationMove) obj;
            return sourceColumnIndex == move.sourceColumnIndex && card.equals(move.card);
        } else {
            return false;
        }
    }
    
    @Override
    public String toString() {
        return "Move " + card + " from column " + sourceColumnIndex + " to foundation.";
    }
    
}

File: cpp/Solitaire/Klondike/silnith/game/solitaire/move/ColumnToFoundationMove.cpp
#include <silnith/game/solitaire/move/ColumnToFoundationMove.h>

#include <stdexcept>

using namespace silnith::game::deck;

using namespace std;
using namespace std::literals::string_literals;

namespace silnith::game::solitaire::move
{
    vector<shared_ptr<solitaire_move>> ColumnToFoundationMove::find_moves(board const& b)
    {
        vector<shared_ptr<solitaire_move>> moves{};
        for (size_t index{ 0 }; index < board::num_columns; index++)
        {
            column const& source_column{ b.get_column(index) };
            if (source_column.has_face_up_cards())
            {
                card const& top_card{ source_column.get_top_card() };
                if (b.can_add_to_foundation(top_card))
                {
                    moves.emplace_back(make_shared<ColumnToFoundationMove>(index, top_card));
                }
            }
        }
        return moves;
    }

    ColumnToFoundationMove::ColumnToFoundationMove(size_t source_column_index, card const& c)
        : source_column_index{ source_column_index },
        _card{ c }
    {
        if (source_column_index > board::num_columns)
        {
            throw out_of_range("Source column "s + std::to_string(source_column_index) + " is out of range."s);
        }
    }

    ColumnToFoundationMove::ColumnToFoundationMove(size_t source_column_index, board const& b)
        : ColumnToFoundationMove(source_column_index, b.get_column(source_column_index).get_top_card())
    {}

    ColumnToFoundationMove::ColumnToFoundationMove(size_t source_column_index, shared_ptr<board> const& b)
        : ColumnToFoundationMove(source_column_index, *b)
    {}

    bool ColumnToFoundationMove::has_cards(void) const
    {
        return true;
    }

    vector<card> ColumnToFoundationMove::get_cards(void) const
    {
        return vector<card>{ _card };
    }

    bool ColumnToFoundationMove::is_stock_pile_modification(void) const
    {
        return false;
    }

    bool ColumnToFoundationMove::is_stock_pile_advance(void) const
    {
        return false;
    }

    bool ColumnToFoundationMove::is_stock_pile_recycle(void) const
    {
        return false;
    }

    bool ColumnToFoundationMove::is_from_stock_pile(void) const
    {
        return false;
    }

    bool ColumnToFoundationMove::is_from_foundation(void) const
    {
        return false;
    }

    bool ColumnToFoundationMove::is_from_column(void) const
    {
        return true;
    }

    bool ColumnToFoundationMove::is_from_column(size_t column_index) const
    {
        return column_index == source_column_index;
    }

    size_t ColumnToFoundationMove::get_source_column_index(void) const
    {
        return source_column_index;
    }

    bool ColumnToFoundationMove::is_to_foundation(void) const
    {
        return true;
    }

    bool ColumnToFoundationMove::is_to_column(void) const
    {
        return false;
    }

    bool ColumnToFoundationMove::is_to_column(size_t column_index) const
    {
        return false;
    }

    size_t ColumnToFoundationMove::get_destination_column_index(void) const
    {
        throw invalid_argument("Not a move to a column.");
    }

    shared_ptr<board const> ColumnToFoundationMove::apply(shared_ptr<board const> const& b) const
    {
        column const& source_column{ b->get_column(source_column_index) };
        pair<card, column> pair{ source_column.extract_card() };
        card moved_card{ pair.first };
        column new_column{ pair.second };

        vector<column> new_columns{};
        for (size_t index{ 0 }; index < board::num_columns; index++)
        {
            if (index == source_column_index)
            {
                new_columns.emplace_back(new_column);
            }
            else
            {
                new_columns.emplace_back(b->get_column(index));
            }
        }

        vector<card> const& stock_pile{ b->get_stock_pile() };
        size_t stock_pile_index{ b->get_stock_pile_index() };
        map<suit, vector<card>> const& new_foundation{ b->get_foundation_plus_card(moved_card) };

        return make_shared<board>(new_columns, stock_pile, stock_pile_index, new_foundation);
    }

    bool ColumnToFoundationMove::operator==(ColumnToFoundationMove const& other) const
    {
        return source_column_index == other.source_column_index
            && _card == other._card;
    }

    string to_string(ColumnToFoundationMove const& move)
    {
        return "Move "s
            + to_string(move.get_cards().front())
            + " from column "s
            + std::to_string(move.get_source_column_index())
            + " to foundation."s;
    }

    wstring to_wstring(ColumnToFoundationMove const& move)
    {
        return L"Move "s
            + to_wstring(move.get_cards().front())
            + L" from column "s
            + std::to_wstring(move.get_source_column_index())
            + L" to foundation."s;
    }

    ostream& operator<<(ostream& out, ColumnToFoundationMove const& move)
    {
        out << to_string(move);
        return out;
    }

    wostream& operator<<(wostream& out, ColumnToFoundationMove const& move)
    {
        out << to_wstring(move);
        return out;
    }
}
